salt/cli/support/collector.py

```diff
@@ -0,0 +1,497 @@
+                if not isinstance(action[action_name], six.string_types):
+                    info, output, conf = self._get_action(action)
+                    action_type = self._get_action_type(action)  # run:<something> for runners
+                    if action_type == self.RUNNER_TYPE:
+                        self.out.put('Running {}'.format(info.lower()), indent=2)
+                        self.collector.write(info, self._local_run(conf), output=output)
+                    elif action_type == self.CALL_TYPE:
+                        if not conf.get('salt.int.intfunc'):
+                            self.out.put('Collecting {}'.format(info.lower()), indent=2)
+                            self.collector.write(info, self._local_call(conf), output=output)
+                        else:
+                            self.collector.discard_current()
+                            self._internal_function_call(conf)
+                    else:
+                        self.out.error('Unknown action type "{}" for action: {}'.format(action_type, action))
+                else:
+                    # TODO: This needs to be moved then to the utils.
+                    #       But the code is not yet there (other PRs)
+                    self.out.msg('\n'.join(salt.cli.support.console.wrap(action[action_name])), ident=2)
+
+    def _get_action_type(self, action):
+        '''
+        Get action type.
+        :param action:
+        :return:
+        '''
+        action_name = next(iter(action or {'': None}))
+        if ':' not in action_name:
+            action_name = '{}:{}'.format(self.CALL_TYPE, action_name)
+
+        return action_name.split(':')[0] or None
+
+    def collect_targets_data(self):
+        '''
+        Collects minion targets data
+        :return:
+        '''
+        # TODO: remote collector?
+
+    def _cleanup(self):
+        '''
+        Cleanup if crash/exception
+        :return:
+        '''
+        if (hasattr(self, 'config')
+           and self.config.get('support_archive')
+           and os.path.exists(self.config['support_archive'])):
+            self.out.warning('Terminated earlier, cleaning up')
+            os.unlink(self.config['support_archive'])
+
+    def _check_existing_archive(self):
+        '''
+        Check if archive exists or not. If exists and --force was not specified,
+        bail out. Otherwise remove it and move on.
+
+        :return:
+        '''
+        if os.path.exists(self.config['support_archive']):
+            if self.config['support_archive_force_overwrite']:
+                self.out.warning('Overwriting existing archive: {}'.format(self.config['support_archive']))
+                os.unlink(self.config['support_archive'])
+                ret = True
+            else:
+                self.out.warning('File {} already exists.'.format(self.config['support_archive']))
+                ret = False
+        else:
+            ret = True
+
+        return ret
+
+    def run(self):
```

Comment:
Can we factor what's in here to call some smaller functions that are easier to understand?